### PR TITLE
Load authentication information for all registries in provided secrets

### PR DESCRIPTION
**Pull request by JoshBroomberg**
Nvidia commitments require pulling images from both quay and an authenticated NGC registry. In the V1 builder, this is trivial as credentials for both registries can be inserted into the `domino-quay` secret. But, currently, Forge loads only the specific registry and not all the registries in this secret.

Changing the custom resource created by Nucleus to load the NGc credentials explicityly is non-trivial because, in most deployments, the NGC host is unauthenticated. This means a naive check for the ngc host in nucleus will attempt to load credentials via the forge CR when none are present or necessary. Considered work arounds were a) central config values configuring when to treat a host as needing authentication or b) removing the forge behavior that fails a build when credentials are specified in the CR but not found in the secret. (a) was deemed undesirably complex from a user config perspective and (b) from a Forge usability perspective (the assert is desired).

This PR solves the problem by replicating V1 behavior in forge, loading all registry configs from an secret that appears in the build CR. Registries not explicitly listed in the CR are loaded with strict lower priority and are overridden by any explicit entries for the same host.

The plan is to backport the version of forge with this fix to 5.0 for use with specific deploys of Domino. This fix will then ship in 5.1 alongside other work to bring NGC images into Domino more formally. In future releases, we will add better handling of registry credentials to nucleus so that it knows which creds it needs and where to find them. We will then 'revert' the logic in this PR. 

- Changes the authentication configuration logic to load all the entries in a dockerconfig.json type secret.
  - Implicit configs are assumed SSL. An explicit entry for the config must be made to use non SSL registries.
  - Explicit configs take strict precedence, overriding implicit configs.

- Preserves the assert functionality which provides a clear error when the specified registry is *missing* from the secret

- Adds additional logging during configuration to help users debug issues with authentication config

### CODE REVIEW

**Comment by fraenkel** [comment]
Where are the tests for these changes?

**Comment by JoshBroomberg** [comment]
Marking draft again while I work on tests and a more intelligent SSL inference scheme

**Comment by JoshBroomberg** [comment]
> Where are the tests for these changes?

@fraenkel thanks for the bump. I was still working on them. I have now added thorough tests.

**Comment by JoshBroomberg** [comment]
@sonnysideup this is now ready for review

File: internal/credentials/docker.go [hunk 1]
```diff
@@ -16,23 +15,35 @@ type DockerConfigJSON struct {
...
+		authConfigs[host] = types.AuthConfig(conf)
 	}
 
-	ac, ok := cf.GetAuthConfigs()[host]
+	return authConfigs, nil
+}
+
+// ExtractBasicAuthForHost will extract the basic auth info for a registry host from an AuthConfigs instance
+func ExtractBasicAuthForHost(authConfigs AuthConfigs, host string) (string, string, error) {
+	ac, ok := authConfigs[host]
```
Comment (fraenkel) [inline internal/credentials/docker.go:39]: why not flip this around
```suggestion
	if ac, ok := authConfigs[host]; ok {
	  return ac.Username, ac.Password, nil
	}
	...
	return the error
```

File: internal/credentials/docker_test.go [hunk 2]
```diff
@@ -7,24 +7,31 @@ import (
...
 )
 
-func TestExtractDockerAuth(t *testing.T) {
+func TestExtractAuthForHost(t *testing.T) {
 	hostname := "registry.test"
 	input := []byte(`{"auths":{"registry.test":{"username":"steve-o","password":"awesome"}}}`)
 
 	t.Run("success", func(t *testing.T) {
-		username, password, err := ExtractDockerAuth(input, hostname)
+		authConfigs, err := ExtractAuthConfigs(input)
```
Comment (fraenkel) [inline internal/credentials/docker_test.go:15]: why not just build the authConfigs?

File: internal/credentials/docker_test.go [hunk 3]
```diff
@@ -7,24 +7,31 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestExtractDockerAuth(t *testing.T) {
+func TestExtractAuthForHost(t *testing.T) {
```
Comment (fraenkel) [inline internal/credentials/docker_test.go:10]: You have ExtractAuthConfigs and ExtractBasicAuthForHost but not ExtractAuthForHost.
Either way you are missing tests.

**Review by fraenkel — CHANGES_REQUESTED** [review]
Still missing tests for multiple registries, duplicate registries, overriding registries.

File: internal/buildjob/buildjob.go [hunk 4]
```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
...
 
-	return configs, nil
+	return registryConfigs, nil
 }
 
-func (j *Job) getDockerAuthFromFS(host string) (string, string, error) {
+func (j *Job) getDockerAuthsFromFS() (credentials.AuthConfigs, error) {
 	if _, err := os.Stat(config.DynamicCredentialsFilepath); os.IsNotExist(err) {
-		return "", "", errors.Wrap(err, "filesystem docker credentials missing")
+		return nil, errors.Wrap(err, "filesystem docker credentials missing")
```
[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:302]: I believe they are placed into the file system by a sidecar of some kind. I will update the wording here.

File: internal/buildjob/buildjob.go [hunk 5]
```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
...
 	}
 
 	input, err := ioutil.ReadFile(config.DynamicCredentialsFilepath)
 	if err != nil {
-		return "", "", errors.Wrap(err, "cannot read docker credentials")
+		return nil, errors.Wrap(err, "cannot read docker credentials")
```
[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:307]: I think this should be done globally and is out of scope for this PR which uses the existing patterns present in the code. I have opened #101 for this

File: internal/buildjob/buildjob.go [hunk 6]
```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
...
 	}
 
-	username, password, err := credentials.ExtractDockerAuth(input, host)
+	authConfigs, err := credentials.ExtractAuthConfigs(input)
 	if err != nil {
-		return "", "", errors.Wrapf(err, "cannot process filesystem docker auth for host %q", host)
+		return nil, errors.Wrapf(err, "Failed to read docker auth from filesystem")
```
[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:312]: Good call. Updated wording

[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:312]: I have updated the wording of all the errors in this section of the file for consistency

File: internal/buildjob/buildjob.go [hunk 7]
```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
...
+	return authConfigs, nil
 }
 
-func (j *Job) getDockerAuthFromSecret(ctx context.Context, host, name, namespace string) (string, string, error) {
-	secret, err := j.clientk8s.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
+func (j *Job) getDockerAuthsFromSecret(ctx context.Context, secretName string, secretNamespace string) (credentials.AuthConfigs, error) {
+	secret, err := j.clientk8s.CoreV1().Secrets(secretNamespace).Get(ctx, secretName, metav1.GetOptions{})
 	if err != nil {
-		return "", "", errors.Wrap(err, "cannot find registry auth secret")
+		return nil, errors.Wrap(err, "cannot find registry auth secret")
```
[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:321]: FYI this was the original error message (I only updated the first part of the tuple). I have updated it because you're right, but generally I try to minimize changes to existing code as part of any specific PR

[the remark this answers is not in the document]
Comment (fraenkel) [reply internal/buildjob/buildjob.go:321]: Sure. But I find its also the only time you recognize past mistakes. And given it was touched it becomes fair game.

File: internal/buildjob/buildjob.go [hunk 8]
```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
...
 
-	username, password, err := credentials.ExtractDockerAuth(secret.Data[corev1.DockerConfigJsonKey], host)
+	return authConfigs, nil
+}
+
+func (j *Job) getBasicAuthForHost(authConfigs credentials.AuthConfigs, host string) (string, string, error) {
+	username, password, err := credentials.ExtractBasicAuthForHost(authConfigs, host)
 	if err != nil {
-		return "", "", errors.Wrap(err, "cannot process docker auth from secret")
+		return "", "", errors.Wrapf(err, "Failed to extract docker auth for specified host (%q)", host)
```
[the remark this answers is not in the document]
Comment (JoshBroomberg) [reply internal/buildjob/buildjob.go:338]: The parentheses are stylistic and purely meant to provide visual clarity to help pick out the host string from the line of rendered text

[the remark this answers is not in the document]
Comment (fraenkel) [reply internal/buildjob/buildjob.go:338]: Right but (abc) vs ("abc") could just be `specified host, "host"` or just `specificed host, HOST`
normally it would be context on the log rather than text.

File: internal/credentials/docker.go (hunk 1 above)
Comment (JoshBroomberg) [reply internal/credentials/docker.go:39]: FYI this pattern was present in the original code. I have updated to use your suggestion.

File: internal/credentials/docker_test.go (hunk 3 above)
Comment (JoshBroomberg) [reply internal/credentials/docker_test.go]: Just laziness... I have now done this properly.

File: internal/credentials/docker_test.go (hunk 2 above)
Comment (JoshBroomberg) [reply internal/credentials/docker_test.go:19]: Have now done this properly

**Comment by JoshBroomberg** [comment]
@fraenkel I don't think the tests you mention are missing:

This E2E build tests loading multiple (implicit) registries from a single secret: https://github.com/dominodatalab/forge/pull/99/files#diff-101c92c8a7b860d914df1f85da969b55315415cfd0ea0da5829df53689d343c2

This E2E build tests multiple explicit registries, one of which is a duplicate that (must) override an implicit registry in order for the test to pass: https://github.com/dominodatalab/forge/pull/99/files#diff-c07a8dcc32bfc87feb094ba78f482d44295482ecec5881e116f18185feb91b51

**Comment by fraenkel** [comment]
Sorry but e2e (integration) tests are insufficient. We need to have unit tests to provide what you believe will happen rather than rely on some other testing that may or may not actually behave the way you think.

**Comment by JoshBroomberg** [comment]
@fraenkel the only thing I am not sure about here is if I have added the new k8s fake module in the right way. I believe I have and CI seems to work?

File: internal/buildjob/buildjob_test.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+		Host:     secretHost2,
+		Username: secretUsername2,
+		Password: secretPassword2,
+		NonSSL:   false,
+	}
+
+	singleRegistrySecretName := "SecretWithSingleRegistry"
+	singleRegistrySecretNamespace := "SomeNamespace"
+	singleRegistrySecretContent := []byte(
+		fmt.Sprintf(`{"auths":{"%v":{"username":"%v","password":"%v"}}}`, secretHost1, secretUsername1, secretPassword1))
```
Comment (fraenkel) [inline internal/buildjob/buildjob_test.go:76]: a different way would have been to create a nested map and then use json.Encode to convert it.

File: internal/buildjob/buildjob_test.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+		},
+	)
+
+	ctx := context.Background()
+	log := NewLogger()
+	job := Job{
+		log: log,
+	}
+
+	t.Run("bad_basic_auth_both_inline_and_secret", func(t *testing.T) {
```
Comment (fraenkel) [inline internal/buildjob/buildjob_test.go:134]: its not bad per-say.  The test is testing both inline and secret which is expected to fail.
This could be lifted out since it doesn't depend on anything above.

File: internal/buildjob/buildjob_test.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+		apiRegs := []v1alpha1.Registry{twoRegistriesSecretRegistryConfig, inlineRegistryConfig}
+		registries, err := job.buildRegistryConfigs(ctx, apiRegs)
+
+		assert.NoError(t, err)
+		assert.ElementsMatch(t, registries, []config.Registry{secretRegistry1, inlineRegistry})
+
+		// Test order irrelevant to overriding
+		apiRegsReverseOrder := []v1alpha1.Registry{inlineRegistryConfig, twoRegistriesSecretRegistryConfig}
+		registriesReverseOrder, err := job.buildRegistryConfigs(ctx, apiRegsReverseOrder)
```
Comment (fraenkel) [inline internal/buildjob/buildjob_test.go:211]: so this is a bit tricky since I believe you are walking a map to build the registry list which has no order. Go will randomize this walk so it can fail. If assert.ElementsMatch does not care about order, we are good.

**Review by fraenkel — APPROVED** [review]
decide if we need to fix the test or not.
I would have organized the tests slightly different and had them broken out more.
But I am not going to block this over that.

File: internal/buildjob/buildjob_test.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/dominodatalab/forge/api/forge/v1alpha1"
+	"github.com/dominodatalab/forge/internal/config"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	testK8sClient "k8s.io/client-go/kubernetes/fake"
```
Comment (fraenkel) [inline internal/buildjob/buildjob_test.go:13]: I would have gone with fakeClient or just use fake since it doesn't overlap rather than a long name.

File: internal/buildjob/buildjob_test.go (hunk 10 above)
Comment (JoshBroomberg) [reply internal/buildjob/buildjob_test.go:134]: It depends on the Job/ctx/logs etc. But I have fixed the name.

File: internal/buildjob/buildjob_test.go (hunk 11 above)
Comment (fraenkel) [reply internal/buildjob/buildjob_test.go:211]: why call it registriesReverseOrder if you cannot guarantee the order it comes back as?
